### PR TITLE
decouple Geometry from its GeometryResource in ModelCache

### PR DESCRIPTION
**Pull request by AndrewMeadows**
* NO new new features or functional changes -- internal architecture changes only.  In particular the code affected does the following:
* trigger mesh being added to render pipeline as soon as model geometry load completes
* trigger collision shape being added to physics simulation as soon as collision/model geometry load completes

### CODE REVIEW

File: libraries/model-networking/src/model-networking/ModelCache.h [hunk 1]
```diff
@@ -22,52 +22,30 @@
...
-protected:
-    friend class GeometryMappingResource;
-
-    virtual QSharedPointer<Resource> createResource(const QUrl& url, const QSharedPointer<Resource>& fallback,
-        const void* extra);
-
-private:
-    ModelCache();
-    virtual ~ModelCache() = default;
-};
```
Comment (AndrewMeadows) [inline libraries/model-networking/src/model-networking/ModelCache.h]: `ModelCache` class declaration did not get removed -- it moved down to satisfy declaration order problems when trying to use a typedef.

File: libraries/model-networking/src/model-networking/ModelCache.h [hunk 2]
```diff
@@ -118,30 +96,49 @@ class GeometryResource : public Resource, public Geometry {
...
 
 private slots:
     void resourceFinished(bool success);
     void resourceRefreshed();
 
 private:
     GeometryResource::Pointer _resource;
-    Geometry::Pointer _instance { nullptr };
+    Geometry::Pointer& _geometryRef;
+};
```
Comment (AndrewMeadows) [inline libraries/model-networking/src/model-networking/ModelCache.h]: NetworkGeometry is replaced by GeometryResourceWatcher.  The _instance data member is replaced by _geometryRef which is a reference to an actual Geometry::Pointer rather than a copy.

The `Model` class gets a `GeometryResourceWatcher` for each `Geometry` that it may have to load (e.g. "visible" and "collision").  This give the `Model` class more access to its geometry and makes it possible to generate geometry via code logic -- without it necessarily connected to a valid resource (file or URL).

File: libraries/render-utils/src/Model.cpp [hunk 3]
```diff
@@ -78,6 +78,10 @@ void initCollisionHullMaterials() {
 
 Model::Model(RigPointer rig, QObject* parent) :
     QObject(parent),
+    _renderGeometry(),
+    _collisionGeometry(),
+    _renderWatcher(_renderGeometry),
+    _collisionWatcher(_collisionGeometry),
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/Model.cpp]: Each GeometryResourceWatcher instance must be initialized with a corresponding Geometry::Pointer (for which it watches resources, as the case may be).

File: libraries/render-utils/src/Model.cpp [hunk 4]
```diff
@@ -1248,23 +1247,25 @@ void Model::segregateMeshGroups() {
...
-                _collisionRenderItemsSet << std::make_shared<MeshPartPayload>(networkMesh, partIndex, _collisionHullMaterials[partIndex % NUM_COLLISION_HULL_COLORS], transform, offset);
-            } else {
-                _modelMeshRenderItemsSet << std::make_shared<ModelMeshPartPayload>(this, i, partIndex, shapeID, transform, offset);
-            }
 
-            shapeID++;
+                shapeID++;
+            }
         }
     }
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/Model.cpp]: The interesting change in this context is that we're removing a dependency on FBXGeometry.  Instead we use the corresponding Geometry classes.

File: libraries/render-utils/src/Model.h [hunk 5]
```diff
@@ -263,7 +263,11 @@ class Model : public QObject, public std::enable_shared_from_this<Model> {
     /// \return true if joint exists
     bool getJointPosition(int jointIndex, glm::vec3& position) const;
 
-    NetworkGeometry::Pointer _geometry;
+    Geometry::Pointer _renderGeometry; // only ever set by its watcher
+    Geometry::Pointer _collisionGeometry; // only ever set by its watcher
+
+    GeometryResourceWatcher _renderWatcher;
+    GeometryResourceWatcher _collisionWatcher;
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/Model.h]: As mentioned higher up: Model class gets direct ownership of its Geometry::Pointer instances, and for each there is a watcher which will fire appropriate signals when GeometryResources are done loading.

**Comment by AndrewMeadows** [comment]
To test:

(1) Make sure models for entities properly load and are rendered:
(1a) existing model entities should load and render at login
(1b) newly created model entities should render
(1c) deleting entities should make them vanish

(2) Make sure entity collision shapes load and collide as normal:
(2a) existing entities should collide correctly on login
(2b) newly created model entities with collision shapes set should collide
(2c) deleting entities should make their collision shapes also vanish from physics simulation

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8e2f22d09b5707d9419d9ab32a00a5baa9056fe5 available at: https://gist.github.com/f8680795eccedb08cb75203f0c46b515<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205.dmg

**Comment by AndrewMeadows** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8e2f22d09b5707d9419d9ab32a00a5baa9056fe5 available at: https://gist.github.com/f8680795eccedb08cb75203f0c46b515<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205.dmg

**Comment by AndrewMeadows** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8e2f22d09b5707d9419d9ab32a00a5baa9056fe5 available at: https://gist.github.com/f8680795eccedb08cb75203f0c46b515<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205.dmg

File: libraries/entities-renderer/src/RenderableModelEntityItem.cpp [hunk 6]
```diff
@@ -808,12 +812,13 @@ void RenderableModelEntityItem::computeShapeInfo(ShapeInfo& info) {
                 scaleToFit[i] = 1.0f;
             }
         }
-        for (int i = 0; i < points.size(); ++i) {
-            points[i] = (points[i] * scaleToFit);
+        for (auto points : pointCollection) {
```
Comment (zzmp) [inline libraries/entities-renderer/src/RenderableModelEntityItem.cpp]: Is this a ref? Shouldn't it be `auto&`?

File: libraries/model-networking/src/model-networking/ModelCache.cpp [hunk 7]
```diff
@@ -90,8 +90,8 @@ void GeometryMappingResource::downloadFinished(const QByteArray& data) {
 
 void GeometryMappingResource::onGeometryMappingLoaded(bool success) {
     if (success && _geometryResource) {
-        _geometry = _geometryResource->_geometry;
```
Comment (zzmp) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: I named away form `fbx` because at this point, the FBX geometry should be an internal representation. IMO, FBX should only exist at the interface between reading an FBX and creating an internal representation.

File: libraries/model-networking/src/model-networking/ModelCache.cpp [hunk 8]
```diff
@@ -250,17 +250,15 @@ QSharedPointer<Resource> ModelCache::createResource(const QUrl& url, const QShar
     return QSharedPointer<Resource>(resource, &Resource::deleter);
 }
 
-std::shared_ptr<NetworkGeometry> ModelCache::getGeometry(const QUrl& url, const QVariantHash& mapping, const QUrl& textureBaseUrl) {
+GeometryResource::Pointer ModelCache::fetchResource(const QUrl& url, const QVariantHash& mapping, const QUrl& textureBaseUrl) {
```
Comment (zzmp) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: Why did you change this name?

File: libraries/model-networking/src/model-networking/ModelCache.h (hunk 2 above)
Comment (zzmp) [inline libraries/model-networking/src/model-networking/ModelCache.h]: This breaks naming conventions with other caches.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4da9a9902550cbc1ef7717962f56ebc9bd23ef33 available at: https://gist.github.com/f8680795eccedb08cb75203f0c46b515<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205.dmg

File: libraries/model-networking/src/model-networking/ModelCache.cpp (hunk 7 above)
Comment (AndrewMeadows) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: This name change is actually at the Geometry class level: it went from Geometry::_geometry to Geometry::_fbxGeometry.

I like succinct variable names too, but in this particular case the ambiguity of the simple name is worse than violating the ideal.  In other words: the internal variable names should make the code easier to read and in this case the easier name is a more descriptive one.

Suppose we hired a new developer who knows nothing about our code yet and I were to offer two sentences to this newbie:

"We have a class called Geometry and it has a data member named _geometry."

"We have a class called Geometry and it has a data member named _fbxGeometry."

One of these sentences conveys much more information at very little additional overhead.

File: libraries/model-networking/src/model-networking/ModelCache.cpp (hunk 8 above)
Comment (AndrewMeadows) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: I originally implemented a second simultaneous method to do what I needed while leaving the original in place.  Eventually I removed the first and kept the second.  The new name made more sense to me so I kept it.  I'm willing to change this to be `getResource()`.

File: libraries/model-networking/src/model-networking/ModelCache.h (hunk 2 above)
Comment (AndrewMeadows) [inline libraries/model-networking/src/model-networking/ModelCache.h]: I assume you're talking about these classes:
```
class NetworkTexture : public Resource, public Texture
class NetworkShader : public Resource
```
... and not these:
```
class GeometryResource : public Resource, public Geometry
class Buffer : public Resource
class Texture : public Resource
class Sound : public Resource
class Animation : public Resource
class NetworkClipLoader : public Resource
```
(That's odd... why does NetworkTexture derive from Resource twice?)

My opinion is that those two classes are named wrong but their time for overhaul has not yet come.  Someday we'll want to be able to generate a texture or a shader during runtime and not have it associated with a resource.  When that happens we'll have to fix them.  To the Glorious Future one step at a time.

**Comment by AndrewMeadows** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4da9a9902550cbc1ef7717962f56ebc9bd23ef33 available at: https://gist.github.com/f8680795eccedb08cb75203f0c46b515<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205.dmg

File: libraries/model-networking/src/model-networking/ModelCache.cpp [hunk 9]
```diff
@@ -200,31 +200,31 @@ void GeometryDefinitionResource::downloadFinished(const QByteArray& data) {
...
-    for (const FBXMaterial& material : _geometry->materials) {
+    for (const FBXMaterial& material : _fbxGeometry->materials) {
         materialIDAtlas[material.materialID] = _materials.size();
         _materials.push_back(std::make_shared<NetworkMaterial>(material, _textureBaseUrl));
     }
 
-    std::shared_ptr<NetworkMeshes> meshes = std::make_shared<NetworkMeshes>();
-    std::shared_ptr<NetworkShapes> shapes = std::make_shared<NetworkShapes>();
+    std::shared_ptr<GeometryMeshes> meshes = std::make_shared<GeometryMeshes>();
+    std::shared_ptr<GeometryMeshParts> shapes = std::make_shared<GeometryMeshParts>();
```
Comment (zzmp) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: Rename this to meshParts.

File: libraries/model-networking/src/model-networking/ModelCache.cpp (hunk 8 above)
Comment (zzmp) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: I think that getX is the standard name used by all of the other caches. If you can find counterexamples thats fine, but I'd prefer to stick with convention. Too many things in the caches are already double or poorly named, and its a pretty confusing read.

File: libraries/model-networking/src/model-networking/ModelCache.cpp [hunk 10]
```diff
@@ -337,8 +335,8 @@ bool Geometry::areTexturesLoaded() const {
 }
 
 const std::shared_ptr<const NetworkMaterial> Geometry::getShapeMaterial(int shapeID) const {
-    if ((shapeID >= 0) && (shapeID < (int)_shapes->size())) {
-        int materialID = _shapes->at(shapeID)->materialID;
+    if ((shapeID >= 0) && (shapeID < (int)_meshParts->size())) {
```
Comment (zzmp) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: s/shapeID/meshPartID

File: libraries/model-networking/src/model-networking/ModelCache.cpp [hunk 11]
```diff
@@ -361,26 +359,40 @@ void GeometryResource::resetTextures() {
...
+        }
+    }
+}
+
+void GeometryResourceWatcher::resourceFinished(bool success) {
     if (success) {
-        _instance = std::make_shared<Geometry>(*_resource);
+        _geometryRef = std::make_shared<Geometry>(*_resource);
     }
-    emit finished(success);
```
Comment (zzmp) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: Why do you no longer emit on finish?

File: libraries/model-networking/src/model-networking/ModelCache.h [hunk 12]
```diff
@@ -118,30 +96,49 @@ class GeometryResource : public Resource, public Geometry {
...
+class GeometryResourceWatcher : public QObject {
     Q_OBJECT
 public:
-    using Pointer = std::shared_ptr<NetworkGeometry>;
+    using Pointer = std::shared_ptr<GeometryResourceWatcher>;
 
-    NetworkGeometry() = delete;
-    NetworkGeometry(const GeometryResource::Pointer& networkGeometry);
+    GeometryResourceWatcher() = delete;
+    GeometryResourceWatcher(Geometry::Pointer& geometryPtr) : _geometryRef(geometryPtr) {}
```
Comment (zzmp) [inline libraries/model-networking/src/model-networking/ModelCache.h]: Is it guaranteed to be loaded at this point? You do not want to start watching right away?

File: libraries/render-utils/src/Model.cpp [hunk 13]
```diff
@@ -826,7 +829,7 @@ void Model::setCollisionModelURL(const QUrl& url) {
         return;
```
Comment (zzmp) [inline libraries/render-utils/src/Model.cpp]: The line above this should mirror line 804, shouldn't it?

File: libraries/model-networking/src/model-networking/ModelCache.cpp (hunk 11 above)
Comment (zzmp) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: You hold a shared ref which is checked for an object, so you don't need to emit to have any other state changed. Got it.

**Comment by zzmp** [comment]
Outside of a few small concerns, this looks great to me. :+1:

File: libraries/render-utils/src/Model.cpp (hunk 13 above)
Comment (AndrewMeadows) [inline libraries/render-utils/src/Model.cpp]: @zzmp, I don't understand.  Could you elaborate?

File: libraries/model-networking/src/model-networking/ModelCache.h (hunk 12 above)
Comment (AndrewMeadows) [inline libraries/model-networking/src/model-networking/ModelCache.h]: No I don't think so... the watch still starts at the same place as before.

We're funneling all GeometryResourceWatcher constructions through this one ctor, which guarantees the _geometryRef is properly set -- the Model and its GeometryResourceWatchers are locked in a logical embrace at the moment of construction.

The watching doesn't actually start until the URL is known: e.g. in the implementation of Model::setURL().

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7fc69d53216b579786d88e34191296319d83a665 available at: https://gist.github.com/f8680795eccedb08cb75203f0c46b515<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205.dmg

File: libraries/render-utils/src/Model.cpp (hunk 13 above)
Comment (zzmp) [inline libraries/render-utils/src/Model.cpp]: I'd meant to say that the check in setCollisionModelURL (L828) should mirror the logic in L804, checking that `_url == url && _xWatcher.getUrl() == url`.

setCollisionModelURL just checks `_collisionUrl == url`, and that is sufficient (and is what it currently is), because it is a different code path, so I'll go from "should" to "could change because they are similar but doesn't need to change".

File: libraries/model-networking/src/model-networking/ModelCache.cpp (hunk 8 above)
Comment (AndrewMeadows) [inline libraries/model-networking/src/model-networking/ModelCache.cpp]: Turns out getResource() was already taken by the base class (protected method) so I changed this one to 'getGeometryResource()'.

**Comment by AndrewMeadows** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4036951f3396ac253b4a0bd77af9f82812f4fb54 available at: https://gist.github.com/f8680795eccedb08cb75203f0c46b515<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205-4036951.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205-4036951.dmg

**Comment by kevinmthomasse** [comment]
## QA Test
**QA Tester**: KevinMThomas
**QA Pass**: **QA Pass** FAILURE
**Additional Test Results + Observations**:  Immediate crash upon loading interface.  Log included.  Upon 2nd loading of interface immediate crash again.  Log files included.
[hifi-log.txt](https://github.com/highfidelity/hifi/files/358323/hifi-log.txt)

**Comment by OmegaHeron** [comment]
Fail for me as well running Windows 10.  Crashes immediately upon connection.  Sent a couple of bugsplats.

**Comment by AndrewMeadows** [comment]
I fixed the crash bug.  Should be ready for test again as soon as the build completes.

**Comment by hifi-gustavo** [comment]
Build console output for commit: ef1f87fd2e5747e6b6081add5c28e60d9151c194 available at: https://gist.github.com/f8680795eccedb08cb75203f0c46b515<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205-ef1f87f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205-ef1f87f.dmg

**Comment by OmegaHeron** [comment]
@kevinmthomasse - this is all yours once they fix jenkins error.  I was just looking at it on Linux for another reason.

**Comment by AndrewMeadows** [comment]
test this please

**Comment by OmegaHeron** [comment]
Linux is bailing on Interface compile (both here and in jenkins attempt to build)

`In file included from /usr/local/src/hifi/libraries/model-networking/src/model-networking/ModelCache.cpp:12:0:`
`/usr/local/src/hifi/libraries/model-networking/src/model-networking/ModelCache.h:109:10: error: extra qualification ‘GeometryResourc   eWatcher::’ on member ‘getURL’ [-fpermissive]
     QUrl GeometryResourceWatcher::getURL() const { return (bool)_resource ? _resource->getURL() : QUrl(); }`
`          ^`

**Comment by hifi-gustavo** [comment]
Build console output for commit: ef1f87fd2e5747e6b6081add5c28e60d9151c194 available at: https://gist.github.com/f8680795eccedb08cb75203f0c46b515<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205-ef1f87f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205-ef1f87f.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5619a3e3553240995f6a90135bfa4519340dced8 available at: https://gist.github.com/f8680795eccedb08cb75203f0c46b515<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205-5619a3e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205-5619a3e.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: cba49be9eaa64b28a0af35df60114058dbad115b available at: https://gist.github.com/f8680795eccedb08cb75203f0c46b515<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205-cba49be.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8205-cba49be.dmg

**Comment by RyanDowne** [comment]
test this please

**Comment by AndrewMeadows** [comment]
test this please

**Comment by RyanDowne** [comment]
Tested this PR in Forum

(1) Make sure models for entities properly load and are rendered:  :white_check_mark:
(1a) existing model entities should load and render at login  :white_check_mark:
(1b) newly created model entities should render  :white_check_mark:
(1c) deleting entities should make them vanish  :white_check_mark:

(2) Make sure entity collision shapes load and collide as normal  :white_check_mark:
(2a) existing entities should collide correctly on login  :white_check_mark:
(2b) newly created model entities with collision shapes set should collide :white_check_mark:
(2c) deleting entities should make their collision shapes also vanish from physics simulation (there is no mysterious invisible collision shape where they used to be) :white_check_mark:
